docs/design/020-distributed-execution-and-test-suites.md

```diff
@@ -0,0 +1,257 @@
+
+However, I think that the benefits of such an approach would be outweighed by the drawbacks in terms of imlpementation and deployment complexity. In short, it can probably work, but IMO the UX will be worse. Even if we ignore the complications of deploying multiple software stacks (vs. just the k6 binary), it'd be difficult to implement great error handling and timely handling of things like `abortOnFail` thresholds and `test.abort()` API calls. It can certainly be done, but it'd be a ton of effort to not get a flaky experience...
+
+And, as I explained in the "High Availability" section above, a k6 test failing is not the worst thing that can happen. An inconsistent or a flaky test is a worse user experience in my mind.
+
+### Review of "Missing Parts in the Current State"
+
+Let's get back to my actual proposal of a `execution.Controller` API + new `k6 coordinator` and `k6 agent` sub-commands.
+
+To show that the proposed solution is a viable one, I'll go through the 9 items from the list in the "Missing Parts in the Current State" section and show how the proposal can be used to easily implement them:
+1. `k6 coordinator` assembles the script and its dependencies into a [.tar archive bundle](https://k6.io/docs/misc/archive-command/) and distributes it to all `k6 agent` nodes automatically once they connect to it (either pull-based or push-based).
+2. I showed how `setup()` can be cooperatively execuded only on a single instance and how its result can be distributed to the rest of the instances with the `GetOrCreateData()` pseudo-API described in the "Instance Synchronization API" section.
+3. Similarly, in the "Instance Synchronization API" section I showed how the `SignalAndWait()` pseudo-API can be used to ensure that all instances have finished initializing their VUs and executed `setup()` before they start the actual test execution simultaneously
+4. Most failures will be handled by the coordinator and propaged to waiting instances. For the best UX, instances might subscribe to a special error channel so they can be interupted even during the test execution.
+5. Metric collection was described in the "Metrics Handling" section, and the coordinator will be able to stop the test if a threshold is crossed.
+6. We can probably handle `test.abort()` calls from an instance with a `GetOrCreateData()` call that uses a special `dataID` watched by the coordinator, but it might be simpler to just add a custom API for that... :thinking:
+7. Detecting when the test has finished is trivial with `SignalAndWait()`.
+8. `teardown()` can also be executed the same way we do `setup()`. Though, because it doesn't return any data, it is one of the reasons I think it makes sense to split `GetOrCreateData(id, callback)` into separate `Once(callback)`, `SetData(id)` amd `GetData(id)` calls...
+9. The `k6 coordinator` node can run `handleSummary()` after the test is done, no problems there :tada:
+
```

Comment:
One more thing that would be good to include in this list is:
> 10. termination of k6 processes by coordinator

Currently, k6-cloud relies on `--linger` and sending SIGTERM to k6 processes, in order to have consistent logic around `teardown` and finish of the test. It is another piece of low-level logic which has to be done "externally".

As an illustration. This is actually something that became the main stumbling block in my initial attempt to implement [`setup` and `teardown` run once](https://github.com/grafana/k6-operator/issues/223) in k6-operator. Once `--linger` is set, there is no interface to stop k6 process - it must be killed. There is a [stop test call](https://k6.io/docs/misc/k6-rest-api/#stop-test) in REST API but it doesn't actually terminate the process in case of `--linger`: it only returns the status. So one has to remove the k6 process "manually". That leads to almost identical logic for aborting k6 process as response to user abort VS regular stop of the test run, and so these different cases must be additionally tracked as separate thus complicating the logic even more.

Additionally, Kubernetes space in general "expects" the main processes in pods to exit on their own, and these manual interventions seem almost an interference in a healthy Kubernetes flow. From both design and reliability perspective, it'd be beneficial for k6-operator (and perhaps k6-cloud) to rely on a straight-forward interface of a "k6 coordinator" for these things, e.g.: coordinator executes `teardown` and then sends termination signal to all k6 runners. After that, coordinator exits itself, and its exit is an indicator of test run being de facto finished.

On a related note, probably coordinator should also take care of abort terminations as well? I.e. if one wants to abort the test run, it should be enough to "ping" coordinator. So there should be a user interface to do that as well.

Reply:
I'm not sure I'm getting the point regarding `--linger`. At the moment, it should be used as a workaround for getting the current architecture working in a distributed mode. But with the new architecture proposed here, `--linger` is not required anymore to distribute the test as it is already covered by the Sync API. `--linger` remains for the specific case where the user wants to force `k6` to wait even if the test is finished.

>Kubernetes space in general "expects" the main processes in pods to exit on their own,
 
Regarding this specific point, I think is a valid case to support on k6 HTTP API. It is something mentioned at point `5` for `--linger` here https://github.com/grafana/k6/issues/2804#issue-1478961481. Feel free to open an issue for it.   

>then sends termination signal to all k6 runners

We are reversing the flow, so the coordinator shouldn't be able to contact directly the agents. The unique way should be to emit a specific event through the Sync API to terminate/abort but it means there isn't a guarantee _to kill_ the agent. Does it work for the Operator/k8s life-cycle? 


>On a related note, probably coordinator should also take care of abort terminations as well? I.e. if one wants to abort the test run, it should be enough to "ping" coordinator. So there should be a user interface to do that as well.

Yep, it sounds like a good idea for a single entrypoint for forcing the abort. It can be done as described via the Sync API.

Reply:
If `--linger` won't be required with native distribution, all the better. I was trying to illustrate precisely that point: that existing workarounds with `--linger` in k6 Cloud are neither easy to implement nor natural in a setting like Kubernetes. IOW, if native distribution will work and terminate without `--linger` and manual interventions, that's basically ideal.

> I think is a valid case to support on k6 HTTP API. It is something mentioned at point 5 for --linger here #2804 (comment). Feel free to open an issue for it.

Done, [here](https://github.com/grafana/k6/issues/3583).

From test distribution viewpoint, HTTP call is still an intervention (something must make that call) in comparison to native distribution. But having that call would be better than current state for sure :+1:

> The unique way should be to emit a specific event through the Sync API to terminate/abort but it means there isn't a guarantee to kill the agent. Does it work for the Operator/k8s life-cycle?

:thinking: What do you think is the worst-case behaviour on the agent's side? I.e. if "terminate event" is emitted and something gets broken, will agent's process be stuck or will it exit with non-zero error code after some timeout? Additionally, if the agent can get stuck, when can that happen?